File: leet/stack.py
'''
def tryy(pushed, popped):
	out = []
	for p in popped:
		print "ccc"
		l = len(out)-1
		while(l>=0):
			print "bbb"
			if(out[l]==p):
	                	out.pop()
			else:
        	        	break
        	        l = len(out)-1
		while (len(pushed)-1>0 and pushed[0]!=p):
			print "aa"
			out.append(pushed[0])
			pushed.pop(0)
        	        l = len(out)-1
       	print out
	if (out==[]):
		return True
	return False

pushed = input()
popped = input()
print tryy(pushed, popped)
'''

def func(pushed, popped):
	push=0
	l = len(pushed)
	while(popped!=[]):
		if(push>=l):
			return False
#		print l, push
#		print pushed[push]
#		print popped,
#		print push, pushed
		if(popped!=[] and pushed[push]!=popped[0]):
			push+=1
#			print push, pushed
		elif(pushed[push]==popped[0]):
			pushed.pop(push)
			if(push!=0):
				push-=1
#			if(pushed<0):
#				pushed=0
			l-=1
			popped.pop(0)
#		print push, pushed
	return True	
	
	
pushed = input()
popped = input()
print func(pushed, popped)

File: leet/stack.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>

struct node{
	int data;
	struct node *next;
	struct node *prev;
};

bool validateStackSequences(int* pushed, int n, int* popped, int m){
//printf("bool\n");
	
	if(n<m)
		return false;
		
	int c=0, i=0, k=-1;
	int stack[n];
//	stack[0] = *pushed;
	while(c<m)
	{
	
//printf("a %d %d %d\n", k, i, c);

		while(*(pushed+i)!=*(popped+c))
		{
			k++;
			stack[k] = *(pushed+i);
//printf("b %d %d\n", stack[k], *(popped+c));
			i++;
		}
		c++;
		i++;
//printf("c %d %d\n", *(pushed+i), *(popped+c));
		while(stack[k]==*(popped+c))
		{
//printf("d %d %d\n", stack[k], *(popped+c));
			k--;
			c++;
		}
	}
//	printf("%d %d\n", m, k);
	if(c==m && k==n-m-1)
		return true;
	return false;
}

int main()
{
	int n, m, k;
	scanf("%d %d", &n, &m);
	int *push, *pop;
	push = (int *)malloc(sizeof(int)*n);
	pop = (int *)malloc(sizeof(int)*m);
	
//printf("a\n");
	scanf("%d", push);
	for(int i=1; i<n; i++)
	{
		scanf(" %d", &k);
		*(push+i) = k;
	}
	scanf("%d", pop);
	for(int i=1; i<m; i++)
		scanf(" %d", (pop+i));
	printf(validateStackSequences(push, n, pop, m)?"True\n":"False\n");
	return 0;
}
